File: frontend/src/types/icon.ts
/**
 * TypeScript type definitions for icons and generation
 */

export enum IconCategory {
  FINANCE_INVESTISSEMENT = "finance_investissement",
  IMMOBILIER = "immobilier",
  VEHICULES = "vehicules",
  METIERS = "metiers",
  OBJETS = "objets",
  LIEUX = "lieux",
  DEVISES = "devises",
  ACTIONS = "actions",
  ETATS = "etats",
  ORGANISMES = "organismes",
  NOURRITURE = "nourriture",
  SPORT = "sport",
  OTHER = "other",
}

export interface Icon {
  id: string;
  name: string;
  category: IconCategory;
  prompt: string;
  animation_prompt?: string;
  image_url: string;
  thumbnail_url?: string;
  tags: string[];
  download_count: number;
  created_at: string;
  updated_at?: string;
}

export interface IconList {
  icons: Icon[];
  total: number;
  page: number;
  page_size: number;
  success: boolean;
}

export enum GenerationStatus {
  PENDING = "pending",
  PROCESSING = "processing",
  EXTRACTING_CONCEPTS = "extracting_concepts",
  GENERATING_IMAGES = "generating_images",
  REMOVING_BACKGROUNDS = "removing_backgrounds",
  UPLOADING = "uploading",
  COMPLETED = "completed",
  FAILED = "failed",
}

export enum ConceptPriority {
  HIGH = "high",
  MEDIUM = "medium",
  LOW = "low",
}

export interface ConceptExtraction {
  name: string;
  category: string;
  priority: ConceptPriority;
  visual_description: string;
  context?: string;
}

export interface TranscriptSegment {
  text: string;
  start: number;
  duration: number;
}

export interface GenerationTask {
  task_id: string;
  status: GenerationStatus;
  progress: number;
  message?: string;
  created_at: string;
  updated_at: string;
  completed_at?: string;
  error?: string;
  extracted_concepts?: ConceptExtraction[];
  generated_icons?: string[];
  transcript?: TranscriptSegment[];
  transcript_text?: string;
  metadata?: Record<string, any>;
}

export interface GenerateConceptRequest {
  concept: string;
  category?: string;
  style?: string;
  size?: string;
  include_animation_prompt?: boolean;
}

export interface GenerateYouTubeRequest {
  youtube_url: string;
  max_concepts?: number;
  min_priority?: ConceptPriority;
  style?: string;
  auto_generate?: boolean;
}

export interface GenerateResponse {
  task_id: string;
  status: GenerationStatus;
  message: string;
  estimated_time_seconds?: number;
}

export const CATEGORY_LABELS: Record<IconCategory, string> = {
  [IconCategory.FINANCE_INVESTISSEMENT]: "Finance & Investissement",
  [IconCategory.IMMOBILIER]: "Immobilier",
  [IconCategory.VEHICULES]: "Véhicules",
  [IconCategory.METIERS]: "Métiers",
  [IconCategory.OBJETS]: "Objets",
  [IconCategory.LIEUX]: "Lieux",
  [IconCategory.DEVISES]: "Devises",
  [IconCategory.ACTIONS]: "Actions",
  [IconCategory.ETATS]: "États",
  [IconCategory.ORGANISMES]: "Organismes",
  [IconCategory.NOURRITURE]: "Nourriture",
  [IconCategory.SPORT]: "Sport",
  [IconCategory.OTHER]: "Autre",
};

File: backend/app/models/icon.py
"""
Icon data models and schemas
"""

from pydantic import BaseModel, Field, HttpUrl
from typing import Optional, List
from datetime import datetime
from enum import Enum


class IconCategory(str, Enum):
    """Icon categories based on concept extraction"""
    FINANCE_INVESTISSEMENT = "finance_investissement"
    IMMOBILIER = "immobilier"
    VEHICULES = "vehicules"
    METIERS = "metiers"
    OBJETS = "objets"
    LIEUX = "lieux"
    DEVISES = "devises"
    ACTIONS = "actions"
    ETATS = "etats"
    ORGANISMES = "organismes"
    NOURRITURE = "nourriture"
    SPORT = "sport"
    OTHER = "other"


class IconBase(BaseModel):
    """Base icon schema"""
    name: str = Field(..., description="Icon name/concept")
    category: IconCategory = Field(..., description="Icon category")
    prompt: str = Field(..., description="Generation prompt used")
    animation_prompt: Optional[str] = Field(None, description="Video animation prompt")
    tags: List[str] = Field(default_factory=list, description="Searchable tags")


class IconCreate(IconBase):
    """Schema for creating a new icon"""
    image_data: Optional[str] = Field(None, description="Base64 encoded image data")


class Icon(IconBase):
    """Full icon model with database fields"""
    id: str = Field(..., description="Unique icon ID")
    image_url: HttpUrl = Field(..., description="URL to icon image")
    thumbnail_url: Optional[HttpUrl] = Field(None, description="URL to thumbnail")
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: Optional[datetime] = None
    download_count: int = Field(default=0, description="Number of downloads")

    class Config:
        from_attributes = True


class IconResponse(BaseModel):
    """API response for single icon"""
    icon: Icon
    success: bool = True
    message: Optional[str] = None


class IconList(BaseModel):
    """API response for icon list"""
    icons: List[Icon]
    total: int
    page: int
    page_size: int
    success: bool = True


class IconDownload(BaseModel):
    """Icon download request"""
    icon_id: str
    size: str = Field(default="original", description="Image size: original, 2k, 1k")

    class Config:
        json_schema_extra = {
            "example": {
                "icon_id": "abc123",
                "size": "2k"
            }
        }
